DEVELOPING.md

```diff
@@ -17,7 +32,131 @@ This project:
+
+You must provide a Mapbox access token to display Mapbox-hosted maps in the `DebugApp` test application. Add your Mapbox token to the `MBXAccessToken` key in `DebugApp/DebugApp/Info.plist`. Alternatively, create a [file at `~/mapbox`](https://docs.mapbox.com/help/troubleshooting/private-access-token-android-and-ios/#ios) containing your access token.
+
+### CircleCI
+
+CircleCI's [YAML](https://en.wikipedia.org/wiki/YAML) format can be picky. You can run `make validate` to ensure you've formatted the changes to the config file correctly. This make target will install the [CircleCI command line interface](https://circleci.com/docs/2.0/local-cli/) to validate `.circleci/config.yml`, and run `circleci config validate`
+
+# Testing
+
+## Unit Testing on devices
+
+To run device tests there are few options:
+
+1. Run via Xcode:
+
+  - Install [xcodegen](https://github.com/yonaskolb/XcodeGen).
+  - Run `$ xcodegen` in the root of the repo.
+  - Open the resulting `MapboxMaps.xcodeproj`.
+  - Test (Cmd-U) the `MapboxTestHost` scheme.
+
+2. Trigger via CI by adding `[run device tests]` to a git commit message. These tests also run "nightly".
+(Examples can also be run as tests on CI by adding `[run app device tests]`.)
+
+3. Trigger tests on AWS Device Farm from the command line. Note that this requires certain environment variables to be set; please see the makefile for these.
+`make test-with-device-farm SCHEME=MapboxTestHost APP_NAME=MapboxTestHost`.
+
+4. Trigger tests on a local device (connected by USB) using the same setup as Device Farm testing by running:
+`make local-test-with-device-farm-ipa SCHEME=MapboxTestHost CONFIGURATION=Release ENABLE_CODE_SIGNING=1`
+
+## Integration Tests
+
+Integration tests typically test the integration between components. As such most will require a `MapView`. These can be run locally and on AWS Device Farm.
+
+Since these tests require a map view, they also need a `UIWindow` and Metal rendering to work. For both testing scenarios, if there's no valid Metal device the test will be skipped (i.e. not failure/success).
+
+### What happens if I run from `MapboxTestHost`?
+
+There **is** a host application, so `MapViewIntegrationTestCase` fetches the existing window and view controller before adding the MapView to it.
+
+- These tests can be run on devices.
+- Tests will be skipped on simulators < iOS 13 because there's no valid Metal device. (iOS 13+ has a simulated Metal device.)
+
+### What happens if I run from `MapboxMaps`
+
+There is **no host application**, so `MapViewIntegrationTestCase` creates its own `UIWindow` and root view controller, before adding the MapView to it.
+
+- These tests cannot be run on devices.
+- Tests will be skipped on simulators < iOS 13 because there's no valid Metal device.
+- Tests will be skipped entirely on CircleCI because of the VM (no Metal). **So for CI purposes, integration tests should be run on AWS Device Farm via `MapboxTestHost`**
```

Comment:
We should add a note somewhere about *when* to run integration tests in CI since it's opt-in. @OdNairy @evil159 any opinions of when this should happen? Should we require it for PR merge, but agree to typically only run them after code review? Or just keep them optional?

Reply:
They run post-merge either way.